packages/internal/src/utils.js

```diff
@@ -348,3 +349,101 @@ export const allValues = async obj => {
   // @ts-expect-error cast
   return harden(fromEntries(zip(keys(obj), resolved)));
 };
+
+/**
+ * A tee implementation where all readers are synchronized with each other.
+ * They all consume the source stream in lockstep, and any one returning or
+ * throwing early will affect the others.
+ *
+ * @template [T=unknown]
+ * @param {AsyncIterator<T, void, void>} stream
+ * @param {number} quantity
+ */
+export const syncTee = (stream, quantity) => {
+  /** @type {IteratorReturnResult<void> | undefined} */
+  let doneResult;
+
+  /** @typedef {IteratorResult<(value: PromiseLike<IteratorResult<T>>) => void>} QueuePayload */
+  /** @type {import('@endo/stream').AsyncQueue<QueuePayload>[]} */
+  const queues = [];
+
+  /** @returns {Promise<void>} */
+  const pullNext = async () => {
+    const requests = await Promise.allSettled(queues.map(queue => queue.get()));
+    const rejections = [];
+    /** @type {Array<(value: PromiseLike<IteratorResult<T>>) => void>} */
+    const resolvers = [];
+    let done = false;
```

Comment:
By "done" we mean "one of the readers stopped reading early", right? Maybe use `readerDone` or `readerStopped` or something similar? Just trying to capture the fruits of my ignorance while I still have it :)

Reply:
`done` is really the term used in `IteratorResults` which have the shape of `{done: boolean, value: any}`.